internal/fs/fs.go

```diff
@@ -2090,20 +2090,27 @@ func (fs *fileSystem) renameHierarchicalDir(ctx context.Context, oldParent inode
 		return err
 	}
 
+	oldDirName := inode.NewDirName(oldParent.Name(), oldName)
+	newDirName := inode.NewDirName(newParent.Name(), newName)
+
 	// If the call for getBucketDirInode fails it means directory does not exist.
 	newDirInode, err := fs.getBucketDirInode(ctx, newParent, newName)
 	if err == nil {
 		// If the directory exists, then check if it is empty or not.
 		if err = fs.checkDirNotEmpty(newDirInode, newName); err != nil {
 			return err
 		}
+
+		// This refers to an empty destination directory.
+		// The RenameFolder API does not allow renaming to an existing empty directory.
+		// To make this work, we delete the empty directory first from gcsfuse and then perform rename.
+		newParent.Lock()
```

Comment:
I feel locking parent should be enough as it will not give permission to access newDirInode itself if parent is lock. Also we did same for non-hns bucket to delete old directory. [ref](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/internal/fs/fs.go#L2219)